### ISSUE TITLE
Replace database table configuration with Settings API?

### ISSUE THREAD

**Issue by kaittodesk** [sendsmaily/sendsmaily-wordpress-plugin#70]
I am thinking of replacing custom database table based configuration options storing with [Wordpress' Settings API](https://developer.wordpress.org/plugins/settings/settings-api/), maybe we can utilize it to provide a better plugin experience to the user.

**Comment by TomAbel** [comment]
Settings API would allow us to ditch the plugin's database tables and use [WordPress' own option table](https://developer.wordpress.org/apis/handbook/options/). Settings API would allow us to ditch the admin template files as well, we would only need to specify input fields in HTML.

**Comment by kaittodesk** [comment]
I am not 100% certain Settings API can be used as we would like. From the information I have gathered Settings API can be used to add settings fields "Settings" menu item in WordPress' admin panel. I am not sure about adding subpages to "Settings" menu item, because `register_setting` states about `$option_group`:

> Default allowed option key names include 'general', 'discussion', 'media', 'reading', 'writing', 'misc', 'options', and 'privacy'.

Using the subpage option wouldn't be ideal either, because it can make finding Smaily settings harder.

@TomAbel, have you taken an in-depth look of the Settings API (or tried it out)? Would it be possible to keep the menu item placement as it is, but switch forms to Settings API?

If Settings API isn't as flexible as I was hoping, then I think the only option would be to use only Options API.

**Comment by TomAbel** [comment]
> Using the subpage option wouldn't be ideal either, because it can make finding Smaily settings harder.

It can still be used with the `admin_menu` hook, so it appears on the left sidebar.

**Comment by kaittodesk** [comment]
OK, it is good to know that the menu position can remain the same.

But what about `register_setting` states about `$option_group`? Can the list of allowed option key names can be extended?

**Comment by TomAbel** [comment]
> But what about register_setting states about $option_group? Can the list of allowed option key names can be extended?

It's possible to specify the option group and option name as `smailyforwp_api_option`, specifying a different option group would pop an notice/error in my environment.

On a different note, I don't think Settings API is worth it. I can see it being helpful for themes and plugins which have a lot of options that only need sanitization (for example title, name). Our plugin has only a few input fields, the bulk of them for verifying API credentials, of which users need ample feedback on (wrong subdomain, credentials valid).
We would also need to disable auto-loading options. The one way to do that, is to run `update_option` with the `autoload` parameter set to false. This however would void the benefit of having WP handle the option management itself, as we would be changing options. 

However I think we may should implement any benefits of Settings API, like visual consistency and security measures; and use our own implementation of saving/getting data via Options API

**Comment by kaittodesk** [comment]
Yeah, I was also questioning if Settings API would provide any real value to our use case or just make implementing desired user experience more difficult than it should be.

> However, I think we should implement any benefits of Settings API, like visual consistency and security measures; and use our own implementation of saving/getting data via Options API

Can you please elaborate on this a bit or provide a code example (or a pattern) on how the implementation would look like?

**Comment by TomAbel** [comment]
```
smaily_admin_save() {
	validateOp();
	switch ( $form_data['op'] ) {
		case 'validateApiKey':
			$api_credentials = validateAndSanitize( $form_data );
			$request = makeRequest( $api_credentials );
			if ($request) {
				update_option( 'smailyforwp_api_option', $api_credentials, 'no' );
			}
```
```
function smaily_admin_render() {
	$template = new Smaily_Plugin_Template( 'html/admin/page.php' );

	// Load configuration data.
	$data       = get_option('smailyforwp_api_option');
	$template->assign( (array) $data );
```

**Comment by kaittodesk** [comment]
Options API is straight forward and easy to follow, but I am more interested in this part:

> [...] implement any benefits of Settings API, like visual consistency and security measures; [...]

How were you planning to implement visual consistency and security measures?

**Comment by TomAbel** [comment]
Visually, the API credentials section uses styling which WordPress itself uses. But the section after entering credentials does not look anything like the first section. It does not use table rows to keep things aligned.
![image](https://user-images.githubusercontent.com/33309474/106463844-94449680-64a0-11eb-9ebd-73f3f4b47f26.png)

I think something else can be used instead of jQuery tabs to switch between basic and advanced, as its styling can be changed by other plugins that also use jQuery.
